### ISSUE TITLE
Custom Style Rules to Retitle Tabs based on URL & Container?

### ISSUE THREAD

**Issue by viking2** [piroor/treestyletab#2640]
I found an Extension, Tab ReTitle, that let me permanently change the title on the Tabs based on the URL (https://github.com/Lazyuki/ReTitle).  However, it does not let me use different titles for the same URL in different containers.

Is it possible to write custom style rules for TST that changes the Tab Title based on URL _and_ Container?
For example:
Tab for google.com in container1 would change title to Test1
Tab for google.com in container2 would change title to Test2

(I found custom style rules for TST that changes the _appearance_ of the Tab depending on the container.)

**Comment by piroor** [comment]
It looks out of purpose of this project. Container support should be done by ReTitle itself ( https://github.com/Lazyuki/ReTitle/issues/27 ) because people who don't use TST also use it.

**Comment by viking2** [comment]
I understand. However, I just wondered if it is possible to use Custom Style Rules to accomplish this?

**Comment by piroor** [comment]
You can add suffix or prefix like following:

```css
tab-item.contextual-identity-firefox-container-1 tab-label span::after {
  content: "[Personal]"
}
tab-item.contextual-identity-firefox-container-2 tab-label span::after {
  content: "[Work]"
}
tab-item.contextual-identity-firefox-container-3 tab-label span::after {
  content: "[Banking]"
}
tab-item.contextual-identity-firefox-container-4 tab-label span::after {
  content: "[Shopping]"
}
```

But complete replacement is impossible. It is a restriction of CSS spec.

**Comment by irvinm** [comment]
... you could also do this per site you wanted to override and place the title before the actual label.  It would certainly be better if the other addon simply added support for being able to create separate rules per site per container.

```
tab-item[data-current-uri^="https://www.google.com"].contextual-identity-firefox-container-1 tab-label span::before {
  content: "[My Personal Google Title] "
}
tab-item[data-current-uri^="https://www.google.com"].contextual-identity-firefox-container-2 tab-label span::before {
  content: "[My Work Google Title] "
}
tab-item[data-current-uri^="https://www.google.com"].contextual-identity-firefox-container-3 tab-label span::before {
  content: "[My Banking Google Title] "
}
tab-item[data-current-uri^="https://www.google.com"].contextual-identity-firefox-container-4 tab-label span::before {
  content: "[My Shopping Google Title] "
}
```

![image](https://user-images.githubusercontent.com/979729/87223677-e24f4c80-c34c-11ea-97f6-9c8b3612e313.png)

**Comment by viking2** [comment]
Thanks @piroor & @irvinm !

This was exactly what I was looking for :-). With piroor's solution I can combine with Tab Retitle, and with irvinm's solution,  I can replace the original Title with a period or a space.
You should publish theses solutions on your code snippets wiki!

I assume that the container numbers depends on creation order? It will never change?
Is it possible to address by their name instead of number (I tried replacing "1" by "Personal" but it didn't work)?

I hope that https://github.com/Lazyuki/ReTitle/issues/27 will eventually implement this in his addon. It would be a lot more convenient (I have lot of "re-titling" to add in the CSS now).  

BTW, an OT question (I have searched in vain). I have a **lot** of containers. Is there an addon that let me group the containers in the drop-down list (I use the Multi-Account Containers extension)?

**Comment by piroor** [comment]
> I assume that the container numbers depends on creation order? It will never change?

The built-in 4 containers look to have same ID on most runtime environments. Other extra containers will have ID based on their creation order on your environment, and they won't be changed, but I think old IDs are possibly recycled if you delete some containers and create new.

> Is it possible to address by their name instead of number (I tried replacing "1" by "Personal" but it didn't work)?

Sorry but TST doesn't provide such ability for now. The container information is prepared just for TST's built-in indicator of container colors, and I agree it is not user-customization-friendly. I've done a change to improve this point with the commit 10d8c54 and attribute selector like `tab-item[data-contextual-identity-name="Personal"]` will be available on TST 3.5.12 or later.

**Comment by viking2** [comment]
> Sorry but TST doesn't provide such ability for now. The container information is prepared just for TST's built-in indicator of container colors, and I agree it is not user-customization-friendly. I've done a change to improve this point with the commit [10d8c54](https://github.com/piroor/treestyletab/commit/10d8c545fa79fbbba7fc88acc816e49791102349) and attribute selector like `tab-item[data-contextual-identity-name="Personal"]` will be available on TST 3.5.12 or later.

Thanks. Sounds great!

I am also using the Temporary-Containers extension.  The containers show as tmp1, tmp2, tmp3, etc in the address bar.  In TST, they just show as a gray horizontal lines (at Tab right). 
Is there some way, in the Style Rules, to also show these names in TST?

**Comment by piroor** [comment]
On the next version selectors like `tab-item[data-current-uri^="https://www.google.com"][data-contextual-identity-name="tmp1"] tab-label span::before` will be available to do that.
